[SPARK-110] contributing update.

shatteredbeam opened the pull request:
Please note this adds a tagging system for the peer review process, and is open for discussion for ALL team members.  DO contribute to the discussion, and add your input/suggest changes.

Significant changes have been added to the CONTRIBUTING.md file.  Please read it over carefully, as once adopted it will become policy for the project.

You can see the rendered file here: https://github.com/shatteredbeam/pipsqueak3/blob/doc/contributing_update_tagging/CONTRIBUTING.md

@nathanctech @beepbeat

shatteredbeam committed:
contributing update.

shatteredbeam committed:
fixing tag table.

shatteredbeam committed:
fixing tag table...again.  Thanks GH.

314numberpi commented on CONTRIBUTING.md:
```diff
@@ -67,20 +65,52 @@ All pull requests must pass standard checks:
+[Discussion]|Discussion is required, but this is not blocking an approval.
+[Incomplete]|Incomplete code and/or Pull Request.
+[Kudos]|The reviewer liked this, specifically. Go you.
+[Nitpick]|Same as suggestion.  Not blocking, and at authors discretion to change.
```
[Discussion]
```
[Nitpick]|Same as suggestion.  Not blocking, and at authors discretion to change.
```
In what way is this different to [Suggestion] and if it isn't, why have it at all?

shatteredbeam replied:
"This is just a nitpick, but..." has been a really common occurrence in recent peer reviews.  Having this as a tag is just an easy way to say "This _may_ be overly picky, but ..."

TLDR, its an easy way to open the subject regarding something that may be considered pedantic, but still merits attention without going into a full blown [discussion].  Overkill, perhaps, but using it is entirely optional in lieu of a more befitting tag.

314numberpi commented on CONTRIBUTING.md:
```diff
@@ -67,20 +65,52 @@ All pull requests must pass standard checks:
+### Pull Requests Naming
+If working from a registered issue, Include the issue name enclosed in brackets in the title of your Pull Request, ie `[SPARK-99] Update to CONTRIBUTING.md`
+
+Otherwise, use the type of request.
```
[Suggestion]
```
Otherwise, use the type of request.
```
Perhaps in addition to the issue name?

shatteredbeam replied:
I will fix this to something more descriptive. Perhaps "Otherwise, please use the type of request along with the issue name, such as ``[Doc] Updating CONTRIBUTING.md``"

314numberpi replied:
I realise I wasn't clear with my original comment, I mean something along the lines of this, when it is applicable.
```
[Doc] [SPARK-99] Update to CONTRIBUTING.md
```

shatteredbeam replied:
Well, if you're working a 'registered' issue you do have a ticket number.  Otherwise you will not, and have no SPARK identifier to use...unless I still misunderstand?

314numberpi replied:
Your original wording had me thinking of either [SPARK-99] _**or**_ [Doc], dependant on whether a Jira issue exists. Whereas, I believe the [Doc] part should exist regardless of a linked Jira issue.

shatteredbeam replied:
Ahh, I see.  I still think this is overkill, because the type of request is also in the branch name.

ie. ``[SPARK-10298301] Update to CONTRIBUTING.md`` is expected to be in a branch with a ``doc/`` prefix, such as ``doc/update_to_contributing``


Anyone else want to weigh in here?  This is a group decision :)  @Eearslya @theunkn0wn1 @beepbeat

Eearslya replied:
I actually feel like "bugfix/feature/doc" is a task better suited for labels, rather than a tag within the title.

shatteredbeam replied:
I agree with @Eearslya here.

314numberpi reviewed (commented):
Generally, looks good.

Eearslya commented on CONTRIBUTING.md:
```diff
@@ -67,20 +65,52 @@ All pull requests must pass standard checks:
 
 Before opening a PR, rebase onto develop so your PR can be merged fast-forward only, without merge commits.
-The easiest way to do that is to make the main repo a remote using `git remote add upstream git@github.com:fuelrats/pipsqueak` and then running `git pull --rebase upstream develop`.
+The easiest way to do that is to make the main repository a remote using `git remote add upstream git@github.com:fuelrats/pipsqueak` and then running `git pull --rebase upstream develop`.
```
[Blocking] This command points to mecha2's repo, `pipsqueak`. Mecha3's repo is `pipsqueak3`.

Eearslya commented on CONTRIBUTING.md:
```diff
@@ -67,20 +65,52 @@ All pull requests must pass standard checks:
-Everything should be runnable in your dev environment, all necessary configuration files are provided
-in VCS. for pylint you may need to manually point it at `pylint.ini`
+Everything should be executable in your development environment, all necessary configuration files are provided
+in VCS. for pylint you may need to manually point at `pylint.ini`
```
[Blocking] This is now false, as the pylint configuration has moved to the default `.pylintrc`, no "manual pointing" is required.

Eearslya commented on CONTRIBUTING.md:
```diff
@@ -67,20 +65,52 @@ All pull requests must pass standard checks:
+[Discussion]|Discussion is required, but this is not blocking an approval.
+[Incomplete]|Incomplete code and/or Pull Request.
+[Kudos]|The reviewer liked this, specifically. Go you.
+[Nitpick]|Same as suggestion.  Not blocking, and at authors discretion to change.
```
[Nitpick] Grammatical issue: `authors` -> `author's`

Eearslya commented on CONTRIBUTING.md:
```diff
@@ -67,20 +65,52 @@ All pull requests must pass standard checks:
+[Nitpick]|Same as suggestion.  Not blocking, and at authors discretion to change.
+[Style]|Breaking style guidelines, or other issue specifically with code style.
+[Scope]|Out of scope for your Pull Request.
+[Suggestion]|Suggested change.  Not blocking, and at authors discretion to change.
```
[Nitpick] Grammatical issue: `authors` -> `author's`

Eearslya commented on CONTRIBUTING.md:
```diff
@@ -67,20 +65,52 @@ All pull requests must pass standard checks:
-The easiest way to do that is to make the main repo a remote using `git remote add upstream git@github.com:fuelrats/pipsqueak` and then running `git pull --rebase upstream develop`.
+The easiest way to do that is to make the main repository a remote using `git remote add upstream git@github.com:fuelrats/pipsqueak` and then running `git pull --rebase upstream develop`.
+
+## Naming & Commenting Standards
```
[Discussion] Additional standards for commit messages as they pertain to JIRA issues.

theunkn0wn1 commented on CONTRIBUTING.md:
```diff
@@ -67,20 +65,52 @@ All pull requests must pass standard checks:
+[Suggestion]|Suggested change.  Not blocking, and at authors discretion to change.
+[Testing]|Reviewer requests a usage case, or testing to validate. Not blocking.
+[Testing Needed]|Testing or usage case needed before approval. Blocking.
+
```
add a `[error]` block for things that are clearly errors. (whether in methodology or implementation)

theunkn0wn1 commented on CONTRIBUTING.md:
```diff
@@ -67,20 +65,52 @@ All pull requests must pass standard checks:
+
+Adding multiple tags to a comment is fine.  Do not combine tags.
+Good: ``[Suggestion][Kudos] This is great, but...``
+Bad:  ``[Suggestion|Kudos] This is great, but...``
```
[discussion] why not? the latter is easier to read.

shatteredbeam replied:
If we move to using labels, having them separate better facilitates this.  Also,  There are many ways one could adjoin tags, and rather than have to create a separate spec for that, we can just say use them individually.

Eearslya commented on CONTRIBUTING.md:
```diff
@@ -67,20 +65,52 @@ All pull requests must pass standard checks:
+
+Otherwise, use the type of request.
+``[Doc] Update TOS to revoke snickers delivery``
+``[Fix] Mechaclient.py should return None after valid on_message event``
```
[Suggestion] Markdown syntax requires two newlines in order to actually break onto a new line. With just one newline, all three of these lines run together on one.

![https://i.imgur.com/ytTaSSm.png](https://i.imgur.com/ytTaSSm.png)

shatteredbeam replied:
Will fix the typo. Thanks.

Eearslya commented on CONTRIBUTING.md:
```diff
@@ -67,20 +65,52 @@ All pull requests must pass standard checks:
+
+Adding multiple tags to a comment is fine.  Do not combine tags.
+Good: ``[Suggestion][Kudos] This is great, but...``
+Bad:  ``[Suggestion|Kudos] This is great, but...``
```
[Suggestion] Markdown syntax requires two newlines in order to actually break onto a new line. With just one newline, all three of these lines run together on one.

![https://i.imgur.com/fGX0r0D.png](https://i.imgur.com/fGX0r0D.png)

shatteredbeam replied:
Will fix the typo :)

Holladiewal commented on CONTRIBUTING.md:
```diff
@@ -67,20 +65,52 @@ All pull requests must pass standard checks:
+[Suggestion]|Suggested change.  Not blocking, and at authors discretion to change.
+[Testing]|Reviewer requests a usage case, or testing to validate. Not blocking.
+[Testing Needed]|Testing or usage case needed before approval. Blocking.
+
```
[Suggestion] Add a [Questions] tag for things that don't look wrong, but may need a bit more clarification, or are just questions by that person specifically.

shatteredbeam replied:
It may just be me, but I think this is well covered by the [Discussion] tag.

314numberpi replied:
Discussion and questions could cover the same topics. I guess one leans towards an individual response and the other a collective response?

Holladiewal replied:
For me, a Discussion is about how it **should** work, whereas a Question is about how it **does** work. They can indeed overlap.

theunkn0wn1 replied:
For me, I think there is little distinction between what necessitates a discussion and what necessitates a question. As it were, Questions often lead to a discussion as a response.

Thus, while not Discussions are results of questions, I do believe all questions result in discussions.
I think these two terms can be combined without loss in meaning.

shatteredbeam replied:
Is this blocking?  I dont see why we can't have two distinct separate meanings.  One indicates a discussion between the author and the reviewer, the other is a 'Full stop, this needs to be a group decision' tag.

nathanctech replied:
I wonder why we're attempting to define these as blocking/non-blocking. Instead of requiring separate tags, why not make a [Blocking] tag? This way the reviewer can block a merge regardless of other tags on the review.

shatteredbeam replied:
The idea is that some issues are blocking by the very nature of the issue, such as Error, but technically any issue could be blocking, dependent on the reviewer.

I'd like to keep the flexibility in review comments, because we need to make tags available for all reviewers without putting them in a box.

theunkn0wn1 replied:
My concept is any comment is blocking until addressed. If someone raises a question the Pr is blocked until it gets a response or is otherwise resolved. 

The only thing i can see that does not inherently block isa  kudos.

shatteredbeam replied:
That is certainly an way to review things, but it is not the intent of the proposal as discussed in IRC, nor is it the way we have historically reviewed pull requests.  Many times suggestions and nitpicks have been brought up that have explicitly been stated as not blocking.

theunkn0wn1 replied:
Historically it *has* been how pull requests have been treated, unless explicitly enumerated otherwise.

theunkn0wn1 replied:
At the very least, IMHO, suggestions should be acknowledged. at the very least, questions deserve a response (unless marked as resolved by the issuer due to PEBCAK etc)

shatteredbeam replied:
in light of that, can you detail the changes you'd make?  Are you suggesting we invert the system, and all tags are blocking unless we add a [Not Blocking] tag?

theunkn0wn1 replied:
as per conversations in IRC, i believe this conversation is resolved.

shatteredbeam committed:
[SPARK-48] Fixed sample command repo name.
[SPARK-48] Removed reference to pylint.ini, as no manual pointing is required with .pylintrc.
[SPARK-48] Fixed Grammatical issues with tag names.
[SPARK-48] Added [ERROR] Tag
[SPARK-48] Added proposed standards for commit messages.
[SPARK-48] Added missing newlines.

shatteredbeam committed:
[SPARK-48] Fixed a badly formatted line.

nathanctech commented:
I'm liking this as-is! Following it will be a bit of a challenge at first but I'll add it to my cheat sheet for git stuff.

theunkn0wn1 reviewed (dismissed):
I fail to see why `[testing]` is non-blocking. As far as i am concerned, if an extra test case is requested it should either be provided or satifactorally addressed.


as for `[Nitpick]`, i believe it is a superfluous item and should not be in the formal specification, its unnecessary

-- Superficial review, haven't fully read proposal yet

shatteredbeam committed:
[SPARK-48] Separate tags into blocking/non blocking categories
[SPARK-48] Elaborate on blocking changes
[SPARK-48] Add pylint requirements to build pass requirements for pull requests.

shatteredbeam committed:
[SPARK-48] Wording fix for pylint error levels.

theunkn0wn1 reviewed (approved):
May take some adjustment, but im willing to give the current proposal a go.

If needs be, we can adjust this later, but for now it should work.

Approved with reservations

314numberpi reviewed (approved):
Seems to be a marked improvement that can fit most if not all reviewers.
